addons/debugger/module.js

```diff
@@ -0,0 +1,526 @@
+      for (const target of runtime.targets) {
+        for (const soundId in target.sprite.soundBank.soundPlayers) {
+          const soundPlayer = target.sprite.soundBank.soundPlayers[soundId];
+          if (soundPlayer.outputNode) {
+            soundPlayer.outputNode.stop(audioContext.currentTime);
+            soundPlayer._createSource();
+            soundPlayer.outputNode.start(audioContext.currentTime, audioContext.currentTime - soundPlayer.startingUntil + (vm.runtime.currentStepTime / 1000));
+            soundPlayer.startingUntil -= (runtime.currentStepTime / 1000);
+          }
+        }
+      }
+
+      // Move all threads forward one frame in time. For blocks like `wait () seconds`
+      for (const thread of vm.runtime.threads) {
+        if (thread.pauseInfo) {
+          thread.pauseInfo.time += runtime.currentStepTime;
+        }
+      }
+
+    }
+    EVENT_TARGET.dispatchEvent(new CustomEvent(Events.SINGLE_STEPPED));
+  }
+
+  // Returns if we should swap threads.
+  _singleStepThread(thread) {
+    let currentBlockId = thread.peekStack();
+
+    if (!currentBlockId) {
+      thread.popStack();
+
+      if (thread.stack.length === 0) {
+        thread.status = Thread.STATUS_DONE;
+        return true;
+      }
+    }
+
+    const oldActiveThread = this.activeThread;
+    this.activeThread = thread;
+
+    if (thread.target == null) {
+      this.retireThread(thread);
+    } else {
+      this._execute(thread);
+    }
+
+    this.activeThread = oldActiveThread;
+
+    thread.blockGlowInFrame = currentBlockId;
+
+    if (thread.status === Thread.STATUS_YIELD) {
+      thread.status = Thread.STATUS_RUNNING;
+      return true;
+    } else if (thread.status === Thread.STATUS_PROMISE_WAIT || thread.status === Thread.STATUS_YIELD_TICK) {
+      return true;
+    }
+
+    if (thread.peekStack() === currentBlockId) {
+      thread.goToNextBlock();
+    }
+
+    // Look on the stack for more blocks for this thread to execute
+    while (!thread.peekStack()) {
+      thread.popStack();
+
+      if (thread.stack.length === 0) {
+        thread.status = Thread.STATUS_DONE;
+        return !thread.isWarpMode;
+      }
+
+      const stackFrame = thread.peekStackFrame();
+
+      if (stackFrame.isLoop) {
+        if (thread.isWarpMode) {
+          return true;
+        } else {
+          continue;
+        }
+      } else if (stackFrame.waitingReporter) {
+        return true;
+      }
+
+      thread.goToNextBlock();
+    }
+  }
+}
+
+const scratchSequencer = runtime.sequencer;
+runtime.sequencer = new ScratchAddonsSequencer(scratchSequencer);
+// We have to do these because BlockUtility keeps a copy of the original sequencer
+//  and we don't have access to execute's instance of BlockUtility :(
+scratchSequencer.stepToBranch = (...args) => runtime.sequencer.stepToBranch(...args);
+scratchSequencer.stepToProcedure = (...args) => runtime.sequencer.stepToProcedure(...args);
+
+export function isPaused() {
```

Comment:
this would make more sense as a getter

Reply:
I'm not too sure what you mean. I can't just set the `stepToBranch` and `stepToProcedure` methods because than when they are called `this` would refer to the old sequencer.

Reply:
ah, sorry, that was on `isPaused`. It was a variable before and you could easily change it to a getter to change less code